pkg/virt-controller/watch/replicaset.go

```diff
@@ -290,131 +337,224 @@ func (c *VMReplicaSet) scale(rs *virtv1.VirtualMachineReplicaSet, vms []virtv1.V
 	objs, err := c.vmInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
 	if err != nil {
 		return nil, err
 	}
-	vms := []virtv1.VirtualMachine{}
+	vms := []*virtv1.VirtualMachine{}
 	for _, obj := range objs {
-		vms = append(vms, *obj.(*virtv1.VirtualMachine))
+		vms = append(vms, obj.(*virtv1.VirtualMachine))
 	}
 	return vms, nil
 }
 
 // listControllerFromNamespace takes a namespace and returns all VMReplicaSets from the ReplicaSet cache which run in this namespace
-func (c *VMReplicaSet) listControllerFromNamespace(namespace string) ([]virtv1.VirtualMachineReplicaSet, error) {
+func (c *VMReplicaSet) listControllerFromNamespace(namespace string) ([]*virtv1.VirtualMachineReplicaSet, error) {
 	objs, err := c.vmRSInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
 	if err != nil {
 		return nil, err
 	}
-	replicaSets := []virtv1.VirtualMachineReplicaSet{}
+	replicaSets := []*virtv1.VirtualMachineReplicaSet{}
 	for _, obj := range objs {
 		rs := obj.(*virtv1.VirtualMachineReplicaSet)
-		replicaSets = append(replicaSets, *rs)
+		replicaSets = append(replicaSets, rs)
 	}
 	return replicaSets, nil
 }
 
 // getMatchingController returns the first VMReplicaSet which matches the labels of the VM from the listener cache.
 // If there are no matching controllers, a NotFound error is returned.
-func (c *VMReplicaSet) getMatchingController(vm *virtv1.VirtualMachine) (*virtv1.VirtualMachineReplicaSet, error) {
+func (c *VMReplicaSet) getMatchingControllers(vm *virtv1.VirtualMachine) (rss []*virtv1.VirtualMachineReplicaSet) {
 	logger := log.Log
 	controllers, err := c.listControllerFromNamespace(vm.ObjectMeta.Namespace)
 	if err != nil {
-		return nil, err
+		return nil
 	}
 
 	// TODO check owner reference, if we have an existing controller which owns this one
 
 	for _, rs := range controllers {
 		selector, err := v1.LabelSelectorAsSelector(rs.Spec.Selector)
 		if err != nil {
-			logger.Object(&rs).Reason(err).Error("Failed to parse label selector from replicaset.")
+			logger.Object(rs).Reason(err).Error("Failed to parse label selector from replicaset.")
 			continue
 		}
 
 		if selector.Matches(labels.Set(vm.ObjectMeta.Labels)) {
-			// The first matching rs will be returned
-			return &rs, nil
+			rss = append(rss, rs)
 		}
 
 	}
-	return nil, errors.NewNotFound(v1beta1.Resource("virtualmachinereplicaset"), "")
+	return rss
 }
 
-// addVirtualMachine searches for a matching VMReplicaSet, updates it's expectations and wakes it up
+// When a vm is created, enqueue the replica set that manages it and update its expectations.
 func (c *VMReplicaSet) addVirtualMachine(obj interface{}) {
+	vm := obj.(*virtv1.VirtualMachine)
 
-	rsKey := c.getMatchingControllerKey(obj.(*virtv1.VirtualMachine))
-	if rsKey == "" {
+	if vm.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a new vm shows up in a state that
```

Comment:
that's confusing. good catch. Writing our own RS controller has exposed that this logic is trickier than it first appeared.

Reply:
The credit  needs to go to whoever wrote  this for the kubernetes RS :)

Reply:
it's good we're following their design as close as possible then so we get to benefit from their findings.